lib/vrl/DESIGN.md

```diff
@@ -0,0 +1,219 @@
+intended target audience**, which will likely mean choosing different trade-offs
+than you'd make if you were to design the feature for your personal needs.
+
+## Language Limits
+
+There are a number of features that we've so far rejected to implement in the
+language. This might change in the future, but there should be a good reason to
+do so.
+
+- modules (see: [#5507][]
+- classes
+- user-defined functions
+- `goto` statements
+- network calls (see [#4517][])
+
+[#5507]: https://github.com/timberio/vector/issues/5507
+[#4517]: https://github.com/timberio/vector/issues/4517#issuecomment-754160338
+
+## Conventions
+
+### Functions
+
+#### Composition
+
+- Favor function composition over single-purpose functions.
+
+- If a problem needs to be solved in multiple steps, consider adding
+  single-purpose functions for each individual step.
+
+- If useability or readability is hurt by composition, favor single-purpose
+  functions.
+
+#### Naming
+
+- Function names are lower-cased (e.g. `round`).
+
+- Multi-name functions use underscore (`_`) as a separator (e.g.
+  `encode_key_value`).
+
+- Favor explicit verbose names over terse ones (e.g.
+  `parse_aws_cloudwatch_log_subscription_message` over `parse_aws_cwl_msg`).
+
+- Use `parse_*` for string decoding functions (e.g. `parse_json` and
+  `parse_grok`).
+
+- Use `encode_*` for string encoding functions (e.g. `encode_base64`).
+
+- Use `to_*` to convert from one type to another (e.g. `to_bool`).
+
+- Use `format_*` for string formatting functions (e.g. `format_timestamp` and
+  `format_number`).
+
+#### Return Types
+
+- Return boolean from `is_*` functions (e.g. `is_string`).
+
+- Return an error when `parse_*` functions fail to decode the string.
+
+#### Mutability
+
+As a general rule, functions never mutate values.
+
+Favor this:
+
+```coffee
+# explicitly assign the parsed JSON to `.message`
+.message = parse_json(.message)
+```
+
+Over this:
+
+```coffee
+# mutate `.message` in place
+parse_json(.message)
+```
+
+There are exceptions to this rule (such as the `del` function), but they are
+limited, and additional exceptions should be well reasoned.
+
+#### Fallibility
+
+- A function must be marked as fallible if it can fail in any way.
+
+- Design a function with the goal of making it infallible.
+
+- But don't hide fallibility for the sake of persuing the previous rule.
+
+- A function implementation can assume it receives the argument type it has
+  defined.
+
+- `parse_*` functions should almost always error when used incorrectly.
+
+- `to_*` functions must never fail.
+
+#### Signatures
+
+- Functions can have zero or more parameters (e.g. `now()` and `assert(true)`).
+
+- Argument naming follows the same conventions as function naming.
+
+- For one or more parameters, the first parameter must be the "target" of the
+  function (e.g. `parse_regex(target: <string>, pattern: <regex>)`).
+
+- The first parameter must therefor almost always be named `target`.
```

Comment:
Again, not true currently, we (almost always) name the first parameter `value`. When writing this down, it felt like `target` was the better name for it, but that would require updating all functions.

Same as previous comments, we can track this and fix it before 1.0, or keep as is and update this guide to rename `target` to `value`.

Reply:
`target` says to me that the value _will_ be mutated. My personal vote is to keep it `value`.

It looks like the exceptions to using `value` are currently:

- `assert(condition: <boolean>, message: <string>)`
- `assert_eq(left: <any>, right: <any>, message: <string>)
- `del(path: <path>)`
- `exists(path: <path>)`
- `ip_cidr_contains(cidr: <string>, ip: <string>)`
- `ip_subnet(ip: <string>, subnet: <string>)`
- `ip_to_ipv6(ip: <string>)`
- `ipv6_to_ipv4(ip: <string>)`
- `merge(to: <string>, from: <object>, deep: <boolean>)`
- `unnest(path: <path>)`
- `get_env_var(name: <string>)`

Given those examples, I think we could amend this rule slightly to be:

- Call the first parameter `path` if it can only be a VRL path. I think this makes it clearer when paths are required vs. values.
- Otherwise, `value` unless you have a good reason not to (like `merge`, `assert`, and `get_env_var`)

I think we could update the `ip` functions to use `value` as an alias.

Reply:
I'm also pro `value` - though I'd argue sometimes `path` and `value` overlap some...

Reply:
Yeah, you both make a good argument, I'll make the change.